Some Engineer logics on Warheads

FS-21 opened the pull request:
### Engineer logics on Warheads

- Now any `InfantryType`, `VehicleType`, `BuildingType` or `AircraftType` can execute some operations engineers do without loosing the firer in the process.
- `FakeEngineer.CanRepairBridges`, if set to true, when a building with `BridgeRepairHut=yes` linked to a bridge is affected by the Warhead then all destroyed bridge sections will be fixed.
- `FakeEngineer.CanDestroyBridges`, if set to true, when a building with `BridgeRepairHut=yes` linked to a bridge is affected by the Warhead then all the bridge will be destroyed.
- `FakeEngineer.CanCaptureBuildings`, if set to true, a building with `Capturable=true` or `NeedsEngineer=true` is affected by the Warhead then the building will be captured by the house's firer.
- `FakeEngineer.DisarmBombs`, if set to true, an attached bomb will be removed if the target is affected by the Warhead.

In `rulesmd.ini`:
```ini
[SOMEWARHEAD]                          ; WarheadType
FakeEngineer.CanRepairBridges=false    ; boolean
FakeEngineer.CanDestroyBridges=false   ; boolean
FakeEngineer.CanCaptureBuildings=false ; boolean
FakeEngineer.DisarmBombs=false         ; boolean
```
Regarding "Forcing specific weapon against certain targets":
  - `ForceWeapon.Capture` forces specified weapon to be used if the target building is capturable.

In `rulesmd.ini`:
```ini
[SOMETECHNO]   
ForceWeapon.Capture=-1     ; integer, -1 to disable

FS-21 committed:
Initial commit - Engineer logics on Warheads

- Now `infantry` and `units` can execute some operations engineers do without loosing the firer like happens with engineers.
- `FakeEngineer.CanRepairBridges`, if set to true, when a building with `BridgeRepairHut=yes` linked to a bridge is affected by the Warhead then all destroyed bridge sections will be fixed.
- `FakeEngineer.CanDestroyBridges`, if set to true, when a building with `BridgeRepairHut=yes` linked to a bridge is affected by the Warhead then all the bridge will be destroyed.
- `FakeEngineer.CanCaptureBuildings`, if set to true, a building with `Capturable=true` or `NeedsEngineer=true` is affected by the Warhead then the building will be captured by the house's firer.

In `rulesmd.ini`:
```ini
[SOMEWARHEAD]                          ; WarheadType
FakeEngineer.CanRepairBridges=false    ; boolean
FakeEngineer.CanDestroyBridges=false   ; boolean
FakeEngineer.CanCaptureBuildings=false ; boolean
```

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/EngineerLogicsOnWarheads

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/EngineerLogicsOnWarheads

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/EngineerLogicsOnWarheads

FS-21 committed:
Now buildings and aircrafts are supported

FS-21 committed:
Code improvements

Removed all repetitive code related to checks and placed all that in a new method called CanBeAffectedByFakeEngineer(...)

FS-21 committed:
tweak

FS-21 committed:
Merge branch 'develop' into feature/EngineerLogicsOnWarheads

FS-21 committed:
Crash fix & code tweaks

Found a silent crash. Code tweaks

FS-21 committed:
2 new tags and improvements

- Added ForceWeapon.Capture.
- Added FakeEngineer.BombDisarm.
- Code cleanup and improvements.

FS-21 committed:
Merge branch 'develop' into feature/EngineerLogicsOnWarheads

FS-21 committed:
Merge branch 'develop' into feature/EngineerLogicsOnWarheads

TaranDahl commented on src/Ext/Techno/Hooks.ReceiveDamage.cpp:
```diff
@@ -60,6 +61,27 @@ DEFINE_HOOK(0x701900, TechnoClass_ReceiveDamage_Shield, 0x6)
+		&& !pSourceHouse->IsAlliedWith(pTargetHouse)
+		&& (pBuilding->Type->Capturable || pBuilding->Type->NeedsEngineer))
+	{
+		reinterpret_cast<bool(__thiscall*)(BuildingClass*, HouseClass*, bool)>(0x448260)(pBuilding, pSourceHouse, true);
```
pBuilding->SetOwningHouse

NetsuNegi commented:
warhead has its own `BombDisarm`, why do you add `FakeEngineer.DisarmBombs` ?

fagonghaiwo commented:
FakeEngineer.CanDestroyBridges

FS-21 committed:
Destroy bridge fixed

and code improvements like adding the "send event" in the capture building, because I think the radar event the capture function has doesn't trigger the Enter event in map triggers.
Also, now can be triggered both repair bridge & destroy bridge depending of the current bridge status if both tags are enabled in the warhead.

FS-21 committed:
tweak

FS-21 commented:
> warhead has its own `BombDisarm`, why do you add `FakeEngineer.DisarmBombs` ?

Vanilla tag "BombDisarm" makes the wapon unusable for other purposes that isn't defuse bombs, also disrupts the weapons selection if you try to use the multiweapons and that tag isn't in the last weapon of the multiweapon list.
If you use a repairer unit against a damaged unit with an attached bomb say "bye, bye!" to that unit because you won't be able to trigger the BombDisarm weapon because the unit is being repaired by another weapon.

Try yourself if you aren't sure of my findings :-)

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/EngineerLogicsOnWarheads

fagonghaiwo commented:
![bandicam 2025-08-12 19-33-24-499](https://github.com/user-attachments/assets/c6f7780a-fa74-4f1d-9b0e-88c681c8f828)
sad
[FlakGuyATWH]
Versus.bridge=3%
FakeEngineer.CanRepairBridges=1    ; boolean
FakeEngineer.CanDestroyBridges=1   ; boolean
FakeEngineer.CanCaptureBuildings=1 ; boolean
FakeEngineer.DisarmBombs=1         ; boolean

fagonghaiwo commented:
![Uploading bandicam 2025-08-12 19-35-48-152.gif…]()
neither repair

fagonghaiwo commented:
![bandicam 2025-08-13 01-25-45-111](https://github.com/user-attachments/assets/3403b45a-ca99-418b-a62d-f75c579fcccd)
sad again

FS-21 committed:
Merge branch 'develop' into feature/EngineerLogicsOnWarheads

TaranDahl commented on src/Ext/Techno/Body.cpp:
```diff
@@ -632,6 +633,69 @@ AircraftTypeClass* TechnoExt::GetAircraftTypeExtra(AircraftClass* pAircraft)
+	const auto pTypeExt = TechnoExt::ExtMap.Find(pThis)->TypeExtData;
+
+	// Force weapon check
+	int nWeaponIndex = pTypeExt->SelectForceWeapon(pThis, pTarget);
```
Incorrect, this is and should only be called by vanilla hook.

TaranDahl commented on src/Ext/Techno/Body.cpp:
```diff
@@ -632,6 +633,69 @@ AircraftTypeClass* TechnoExt::GetAircraftTypeExtra(AircraftClass* pAircraft)
+	int nWeaponIndex = pTypeExt->SelectForceWeapon(pThis, pTarget);
+
+	if (nWeaponIndex < 0) // Multi weapon check
+		nWeaponIndex = pTypeExt->SelectMultiWeapon(pThis, pTarget);
```
Same

TaranDahl commented on src/Ext/Techno/Hooks.ReceiveDamage.cpp:
It would be better to place these into `DetonateOnOneUnit`.

TaranDahl commented on src/Ext/Techno/Hooks.ReceiveDamage.cpp:
```diff
@@ -60,6 +61,33 @@ DEFINE_HOOK(0x701900, TechnoClass_ReceiveDamage_Shield, 0x6)
+		// Send engineer's "enter" event
+		auto const pTag = pBuilding->AttachedTag;
+		if (args->Attacker && pTag)
+			pTag->RaiseEvent(TriggerEvent::EnteredBy, args->Attacker, CellStruct::Empty);
```
I think this is unreasonable because nothing has entered the target.

TaranDahl replied:
Should raise DestroyedByAnything

FS-21 replied:
Regarding raising TriggerEvent::EnteredBy instead of DestroyedByAnything (or omitting it):

If I'm not wrong in RA2 and YR map scripting (both original campaign missions and custom community maps), mappers use the trigger event Entered By... (Trigger Event 1) on neutral Tech Buildings, mission-critical structures, and Bridge Repair Huts to detect when an engineer captures the building or repairs the bridge.

Since entering the structure was vanilla's only way for an engineer to capture buildings or trigger bridge repairs, EnteredBy serves as the de-facto event hook for map scripts to advance mission objectives (e.g., "Capture Tech Hospital -> Entered by Player -> Objective Completed / Reveal Area").

If we do not raise EnteredBy: Any map trigger expecting the structure to be captured by an engineer will never trigger, breaking campaign progression and custom map scripts.
If we were to raise DestroyedByAnything on capture: It would break missions that have fail conditions like "Protect this structure (Fail if DestroyedByAnybody)", causing an immediate Mission Failed upon capturing the building instead of securing it.
Therefore, raising TriggerEvent::EnteredBy is intentional and necessary to maintain full compatibility with existing map scripts and trigger systems when capturing structures or repairing bridge huts remotely via warheads.

TaranDahl replied:
> If we do not raise EnteredBy: Any map trigger expecting the structure to be captured by an engineer will never trigger, breaking campaign progression and custom map scripts.

Why would an unmerged feature break anything? It doesn't modify any vanilla behavior.

FS-21 replied:
Because is supposed to be used by support units like engineers/saboteurs and those doesn'tdestroy, they capture. I can't imagine these tags in standard units.

TaranDahl commented on src/Ext/Techno/Hooks.TargetEvaluation.cpp:
```diff
@@ -1,7 +1,224 @@
+#pragma region FakeEngineer
+
+// Skipping the next 2 small checks permits the AI to target structures, if used correctly (for example with the compation of AttackFriendlies)
+DEFINE_HOOK(0x6F85C8, TechnoClass_EvaluateObject_RemovingWhatMakesGuardModeAutotargetSelectionUnableToTargetStructures, 0x7)
```
Too long. You should say those in comments.

TaranDahl commented on src/Ext/Techno/Hooks.TargetEvaluation.cpp:
```diff
@@ -1,7 +1,224 @@
+// Skipping the next 2 small checks permits the AI to target structures, if used correctly (for example with the compation of AttackFriendlies)
+DEFINE_HOOK(0x6F85C8, TechnoClass_EvaluateObject_RemovingWhatMakesGuardModeAutotargetSelectionUnableToTargetStructures, 0x7)
+{
+	enum { skipCode = 0x74049F };
```
Should be deleted

TaranDahl commented on src/Ext/WarheadType/Body.cpp:
```diff
@@ -109,6 +111,72 @@ DamageAreaResult WarheadTypeExt::ExtData::DamageAreaWithTarget(const CoordStruct
+	auto const pTag = pBuilding->AttachedTag;
+
+	if (pTag && pOwner)
+		pTag->RaiseEvent(TriggerEvent::EnteredBy, pOwner, CellStruct::Empty);
```
I think this is unreasonable because nothing has entered the target.

TaranDahl commented on src/Ext/WarheadType/Body.h:
```diff
@@ -454,4 +464,5 @@ class WarheadTypeExt
 
 	static void DetonateAt(WarheadTypeClass* pThis, AbstractClass* pTarget, TechnoClass* pOwner, int damage, HouseClass* pFiringHouse = nullptr);
 	static void DetonateAt(WarheadTypeClass* pThis, const CoordStruct& coords, TechnoClass* pOwner, int damage, HouseClass* pFiringHouse = nullptr, AbstractClass* pTarget = nullptr);
+	static void DetonateAtBridgeRepairHut(AbstractClass* pTarget, TechnoClass* pOwner = nullptr, HouseClass* pFiringHouse = nullptr, bool destroyBridge = false);
```
Why is this function declared under WarheadTypeExt but has no relation to WarheadTypeExt or WarheadType?

TaranDahl commented on src/Ext/Techno/Hooks.ReceiveDamage.cpp:
```diff
@@ -60,6 +61,33 @@ DEFINE_HOOK(0x701900, TechnoClass_ReceiveDamage_Shield, 0x6)
+	auto const pBuilding = abstract_cast<BuildingClass*>(pThis);
+
+	// Repair/Destroy bridges at Bridge Repair Huts buildings
+	if (pBuilding && (pWHExt->FakeEngineer_CanRepairBridges || pWHExt->FakeEngineer_CanDestroyBridges))
```
It duplicates the hook in the BuildingClass

TaranDahl commented on src/Ext/Techno/Hooks.TargetEvaluation.cpp:
```diff
@@ -1,7 +1,224 @@
+	const auto pTypeExt = TechnoExt::ExtMap.Find(pFirer)->TypeExtData;
+
+	// Force weapon check
+	int newIndex = pTypeExt->SelectForceWeapon(pFirer, pTarget);
```
Incorrect, this is and should only be called by vanilla hook.

TaranDahl commented on src/Ext/Techno/Hooks.TargetEvaluation.cpp:
```diff
@@ -1,7 +1,224 @@
+	else
+	{
+		// Multi weapon check
+		newIndex = pTypeExt->SelectMultiWeapon(pFirer, pTarget);
```
Same

TaranDahl commented on src/Ext/Techno/Hooks.TargetEvaluation.cpp:
```diff
@@ -1,7 +1,224 @@
+		|| !pBuilding->IsAlive
+		|| pBuilding->Health <= 0
+		|| pFirer->Owner->IsAlliedWith(pVictim)
+		|| (!pBuilding->Type->Capturable && !pBuilding->Type->NeedsEngineer))
```
`NeedsEngineer` is incorrect check.

TaranDahl commented on src/Ext/TechnoType/Body.cpp:
```diff
@@ -64,6 +64,15 @@ int TechnoTypeExt::ExtData::SelectForceWeapon(TechnoClass* pThis, AbstractClass*
+		{
+			if (const auto pBuildingType = abstract_cast<BuildingTypeClass*, true>(pTargetType))
+			{
+				if ((pBuildingType->Capturable || pBuildingType->NeedsEngineer)
```
`NeedsEngineer` is incorrect check.

TaranDahl commented on src/Ext/Techno/Body.cpp:
```diff
@@ -632,6 +633,69 @@ AircraftTypeClass* TechnoExt::GetAircraftTypeExtra(AircraftClass* pAircraft)
+	if (checkCapturableBuilding
+		&& isBuilding
+		&& pWHExt->FakeEngineer_CanCaptureBuildings
+		&& (pBuilding->Type->Capturable || pBuilding->Type->NeedsEngineer)
```
`NeedsEngineer` is incorrect check.

TaranDahl commented on src/Ext/Techno/Hooks.ReceiveDamage.cpp:
```diff
@@ -60,6 +61,33 @@ DEFINE_HOOK(0x701900, TechnoClass_ReceiveDamage_Shield, 0x6)
+	// Capture enemy buildings
+	if (pBuilding && pWHExt->FakeEngineer_CanCaptureBuildings
+		&& !pSourceHouse->IsAlliedWith(pTargetHouse)
+		&& (pBuilding->Type->Capturable || pBuilding->Type->NeedsEngineer))
```
`NeedsEngineer` is incorrect check.

TaranDahl commented on src/Ext/Techno/Hooks.TargetEvaluation.cpp:
```diff
@@ -1,7 +1,224 @@
+	return 0;
+}
+
+DEFINE_HOOK(0x6FCB81, TechnoClass_CanFire_Immune_FakeEngineer_CanCaptureBuildings, 0x5)
```
Can't understand the code of this hook. What does your function have to do with the distance check?

TaranDahl reviewed (changes requested):
Too many problems. I think the author lacks the necessary understanding of the unit's firing procedure.

FS-21 commented:
I applied feedback and small changes.
